File: source/nestjs-petclinic-rest/src/web/rest/vet-rest.controller.ts
import { Controller, Logger, Get, Req, Request, Param, Post, Body, Put, Delete } from '@nestjs/common';
import { VetDTO } from './../../services/dto/vet.dto';
import { VetCreateDTO } from './../../services/dto/vet-create.dto';
import { VetUpdateDTO } from './../../services/dto/vet-update.dto';
import { VetRestService } from './../../services/vet-rest.service';
import { ApiTags, ApiOkResponse, ApiBadRequestResponse, ApiNotFoundResponse, ApiCreatedResponse, ApiNoContentResponse } from '@nestjs/swagger';

@ApiTags('VetRest')
@Controller('api/vets')
export class VetRestController {
  logger = new Logger('VetRestController');

  constructor(private readonly vetRestService: VetRestService) {}

  @Get('')
  @ApiOkResponse({
    description: 'Result retrieved successfully.',
    type: VetDTO,
    isArray: true,
  })
  async getAllVets(@Req() req: Request): Promise<VetDTO[]> {
    const result = await this.vetRestService.getAllVets();
    return result;
  }

  @Get(':vetId')
  @ApiOkResponse({
    description: 'Result retrieved successfully.',
    type: VetDTO,
  })
  @ApiBadRequestResponse({ description: 'Bad request.' })
  @ApiNotFoundResponse({ description: 'Response not found.' })
  async getVet(@Req() req: Request, @Param('vetId') vetId: number): Promise<VetDTO> {
    const result = await this.vetRestService.getVet(vetId);
    return result;
  }

  @Post('')
  @ApiCreatedResponse({
    description: 'The record has been successfully created.',
  })
  @ApiBadRequestResponse({ description: 'Bad request.' })
  async addVet(@Req() req: Request, @Body() dto: VetCreateDTO): Promise<void> {
    return await this.vetRestService.addVet(dto);
  }

  @Put(':vetId')
  @ApiNoContentResponse({
    description: 'Successfully updated.',
  })
  @ApiBadRequestResponse({ description: 'Bad request.' })
  async updateVet(@Req() req: Request, @Param('vetId') vetId: number, @Body() dto: VetUpdateDTO): Promise<void> {
    return await this.vetRestService.updateVet(vetId, dto);
  }

  @Delete(':vetId')
  @ApiOkResponse({
    description: 'Successfully deleted.',
  })
  @ApiBadRequestResponse({ description: 'Bad request.' })
  async deleteVet(@Req() req: Request, @Param('vetId') vetId: number): Promise<void> {
    return await this.vetRestService.deleteVet(vetId);
  }
}

File: source/spring-petclinic-rest/src/main/java/com/spring_petclinic/spring_petclinic_rest/application/rest/VetRestController.java
package com.spring_petclinic.spring_petclinic_rest.application.rest;

import lombok.AllArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import com.spring_petclinic.spring_petclinic_rest.application.models.VetDTO;
import com.spring_petclinic.spring_petclinic_rest.application.models.VetCreateDTO;
import com.spring_petclinic.spring_petclinic_rest.application.models.VetUpdateDTO;
import com.spring_petclinic.spring_petclinic_rest.application.services.VetRestService;
import java.util.List;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import io.swagger.v3.oas.annotations.Parameter;

@RestController
@RequestMapping("/api/vets")
@Tag(name = "VetRestController")
@AllArgsConstructor
public class VetRestController {
    private final VetRestService vetRestService;

    @GetMapping
    @Operation(summary = "getAllVets")
    @ApiResponses(value = {
        @ApiResponse(responseCode = "200", description = "Returns the specified List<VetDTO>.") })
    public ResponseEntity<List<VetDTO>> getAllVets() {
        final List<VetDTO> result = vetRestService.getAllVets();
        if (result.isEmpty()) {
            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
        }
        return new ResponseEntity<>(result, HttpStatus.OK);
    }

    @GetMapping(path = "/{vetId}")
    @Operation(summary = "getVet")
    @ApiResponses(value = {
        @ApiResponse(responseCode = "200", description = "Returns the specified VetDTO."),
        @ApiResponse(responseCode = "400", description = "One or more validation errors have occurred."),
        @ApiResponse(responseCode = "404", description = "Can\'t find an VetDTO with the parameters provided.") })
    public ResponseEntity<VetDTO> getVet(@Parameter(required = true) @PathVariable(value = "vetId") int vetId) {
        final VetDTO result = vetRestService.getVet(vetId);
        if (result == null) {
            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
        }
        return new ResponseEntity<>(result, HttpStatus.OK);
    }

    @ResponseStatus(HttpStatus.OK)
    @PostMapping
    @Operation(summary = "addVet")
    @ApiResponses(value = {
        @ApiResponse(responseCode = "201", description = "Successfully created."),
        @ApiResponse(responseCode = "400", description = "One or more validation errors have occurred.") })
    public void addVet(@Parameter(required = true) @RequestBody VetCreateDTO dto) {
        vetRestService.addVet(dto);
    }

    @ResponseStatus(HttpStatus.OK)
    @PutMapping(path = "/{vetId}")
    @Operation(summary = "updateVet")
    @ApiResponses(value = {
        @ApiResponse(responseCode = "204", description = "Successfully updated."),
        @ApiResponse(responseCode = "400", description = "One or more validation errors have occurred.") })
    public void updateVet(@Parameter(required = true) @PathVariable(value = "vetId") int vetId, @Parameter(required = true) @RequestBody VetUpdateDTO dto) {
        vetRestService.updateVet(vetId, dto);
    }

    @ResponseStatus(HttpStatus.OK)
    @DeleteMapping(path = "/{vetId}")
    @Operation(summary = "deleteVet")
    @ApiResponses(value = {
        @ApiResponse(responseCode = "200", description = "Successfully deleted."),
        @ApiResponse(responseCode = "400", description = "One or more validation errors have occurred.") })
    public void deleteVet(@Parameter(required = true) @PathVariable(value = "vetId") int vetId) {
        vetRestService.deleteVet(vetId);
    }
}